racetrack_module.py

```diff
@@ -0,0 +1,588 @@
+
+    def _update_position(self, j: int, forward: bool, max_flyover: int,
+                         line_from_bottom: int, number_of_lines: int) -> int:
+        """Update position for fly to top algorithm"""
+        if forward:
+            if j + max_flyover > number_of_lines:
+                if j + 1 <= number_of_lines:
+                    return j + 1
+                else:
+                    return j + 1 - max_flyover
+            else:
+                return j + max_flyover
+        else:
+            if j == line_from_bottom:
+                return j + 1
+            elif j - max_flyover < line_from_bottom:
+                return line_from_bottom
+            else:
+                return j - max_flyover
+
+    def _get_save_file_path(self, base_path: str,
+                            sensor_name: str,
+                            flight_altitude: float,
+                            overlap: float,
+                            max_turn: float) -> str:
+        """Get save file path for the waypoint layer"""
+        suggested_file_path, _ = os.path.splitext(base_path)
+        suggested_file_path += (
+            f"_{sensor_name}_{flight_altitude}m_alt_{overlap}overlap_{max_turn}m_turn_racetrack.shp"
+        )
+        file_path, _ = QFileDialog.getSaveFileName(
+            QFileDialog(), "Save Waypoint Layer As", suggested_file_path,
+            "ESRI Shapefile (*.shp *.SHP)"
+        )
+
+        if file_path and not file_path.lower().endswith(".shp"):
+            file_path += ".shp"
+
+        return file_path
+
+    def generate_points_shp_file(self, path_of_line: str, params: FlightParameters,
+                                 crs: QgsCoordinateReferenceSystem) -> Union[QgsVectorLayer, None]:
+        """Generates an SHP-File for the sensor coverage way points"""
+        file_path = self._get_save_file_path(
+            path_of_line, params.sensor_name, params.flight_altitude, params.overlap, params.max_turn_distance
+        )
+
+        if not file_path:
+            return None
+
+        if os.path.exists(file_path):
+            self.iface.messageBar().pushMessage(
+                "Please select a file path that does not already exist",
+                level=Qgis.Warning, duration=4
+            )
+            return None
+
+        return self._create_point_layer(file_path, crs)
+
+    def _create_point_layer(self, file_path: str, crs: QgsCoordinateReferenceSystem) -> Union[QgsVectorLayer, None]:
+        """Create and return a new point layer"""
+        fields = QgsFields()
+        fields.append(QgsField("id", QVariant.Int))
+        fields.append(QgsField("tag", QVariant.String))
+
+        writer = self.layer_utils.create_vector_file_write(
+            file_path, fields, QgsWkbTypes.Point, crs
+        )
+
+        layer = self.iface.addVectorLayer(file_path, "", "ogr")
+        del writer
+        return layer
+
+    def _compute_back_and_forth_points(self, params: dict) -> List[QgsPointXY]:
```

Comment:
Method can be made static
```suggestion
    @staticmethod
    def _compute_back_and_forth_points(params: dict) -> List[QgsPointXY]:
```

Reply:
again a helper function only needed in the class